File: js/766.托普利茨矩阵.js
/**
 * @param {number[][]} matrix
 * @return {boolean}
 */
var isToeplitzMatrix = function(matrix) {
    var judge = (x, y) => {
        var num = matrix[x][y];
        while (x < m && y < n) {
            if (matrix[x][y] !== num) return false;

            x ++, y ++;
        }
        return true;
    };

    var m = matrix.length, n = matrix[0].length;

    for (var i = 0; i < m; i ++) {
        if (!judge(i, 0)) return false;
    }
    for (var i = 0; i < n; i ++) {
        if (!judge(0, i)) return false;
    }
    return true;
};

File: rust/766.托普利茨矩阵.rs
impl Solution {
    pub fn is_toeplitz_matrix(matrix: Vec<Vec<i32>>) -> bool {
        let (m, n) = (matrix.len() as i32, matrix[0].len() as i32);
        for index in 0..m {
            if Solution::judge(index, 0, m, n, &matrix) == false {
                return false;
            }
        }
        for index in 0..n {
            if Solution::judge(0, index, m, n, &matrix) == false {
                return false;
            }
        }
        true
    }

    pub fn judge(x: i32, y: i32, m: i32, n: i32, matrix: &Vec<Vec<i32>>) -> bool {
        let num = matrix[x as usize][y as usize];
        let (mut x, mut y) = (x, y);
        while x < m && y < n {
            if matrix[x as usize][y as usize] == num {
                x += 1;
                y += 1;
            } else {
                return false;
            }
        }
        true
    }
}
